Fix #18464 - Add Redux to statistics page

matanwerbner opened the pull request:
This is a first step to replacing Flux with Redux

Commit:
fixes #18464 - Add Redux to statistics page

This is a first step to replacing Flux with Redux

tbrisker commented:
ok to test

ohadlevy commented:
@gailsteiger please review thanks!

ohadlevy commented on webpack/assets/javascripts/react_app/common/MountingService.js:
```diff
@@ -4,25 +4,25 @@ import PowerStatusContainer from '../components/hosts/PowerStatusContainer';
     PowerStatusContainer: {
       type: PowerStatusContainer,
-      markup: <PowerStatusContainer url={data.url} id={data.id}/>
+      markup: <PowerStatusContainer store={store} url={data.url} id={data.id}/>
```
i assume the other components (besides) statistics should not use the store?

matanwerbner replied:
yup...will fix

ohadlevy commented on webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.js:
```diff
@@ -1,111 +1,43 @@
+import * as StatisticsChartActions from '../../redux/actions/statistics';
 
 class StatisticsChartsList extends React.Component {
   constructor(props) {
```
if the constructor is now noop, can't we remove it?

matanwerbner replied:
yes

ohadlevy commented on webpack/assets/javascripts/react_app/redux/actions/statistics.js:
```diff
@@ -0,0 +1,19 @@
+import API from '../../API';
```
are you still using the "old" api here?

matanwerbner replied:
still using jquery for ajax calls, i figure maybe this should be a seperate issue and discussion.

gailsteiger replied:
+1 for reviewing use of jquery for ajax calls

ohadlevy commented on webpack/assets/javascripts/react_app/redux/actions/statistics.js:
```diff
@@ -0,0 +1,19 @@
+      (jqXHR, textStatus, errorThrown) => dispatch({
+        type: STATISTICS_DATA_FAILURE,
+        payload: errorThrown
+      })
```
what about error?

matanwerbner replied:
error is the line starting with (jqXHR ... 
in then() the first parameter is resolve, the second is reject

ohadlevy commented on webpack/assets/javascripts/react_app/redux/actions/statistics/index.js:
```diff
@@ -0,0 +1,18 @@
+  STATISTICS_DATA_SUCCESS,
+  STATISTICS_DATA_FAILURE
+} from '../../consts';
+
```
is this a duplicate file? whats the diff between actions/statistics and actions/statistics/index.js?

matanwerbner replied:
right, this is the correct one. actions/statistics.js has been removed.

ohadlevy commented on webpack/assets/javascripts/react_app/redux/consts.js:
```diff
@@ -0,0 +1,3 @@
+export const STATISTICS_DATA_REQUEST = 'STATISTICS_DATA_REQUEST';
+export const STATISTICS_DATA_SUCCESS = 'STATISTICS_DATA_SUCCESS';
```
q: does it make sense to group const into one object? e.g.
```js
export const STATISTICS = {
  data: {
    request: '...',
   success: '..'
   ...
}

matanwerbner replied:
this is more common approach for action names...
that way you can explicitly import the action names you need, which makes the code more readable as the reader sees which actions the code will use

gailsteiger replied:
I have a bias towards using only one constants file. Also, prefer a name that doesn't have a spelling error.

matanwerbner replied:
not sure i'm following you :)

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/index.js:
```diff
@@ -0,0 +1,6 @@
+import { combineReducers } from 'redux';
+import statistics from './statistics';
+
+export default combineReducers({
```
is this a global file for all reduces? wont it grow quickly for every component?

matanwerbner replied:
this is only the root of the state tree.. it may grow, but not much, as the leafs may also export "combineReducers"

ohadlevy commented on webpack/assets/javascripts/services/dal/statistics.js:
```diff
@@ -0,0 +1,8 @@
+import axios from 'axios';
```
not used?

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/statistics/index.js:
```diff
@@ -0,0 +1,35 @@
+
+    return state.set('charts', newChartsArray);
+  },
+  [STATISTICS_DATA_FAILURE](state, { payload: { error, id } }) {
```
are these reducers being tested?

matanwerbner replied:
yes, left out in the commit, see below

gailsteiger commented on webpack/assets/javascripts/react_app/redux/actions/statistics.js:
```diff
@@ -0,0 +1,16 @@
+      (jqXHR, textStatus, errorThrown) => dispatch({ type: STATISTICS_DATA_FAILURE, payload: errorThrown })
+    );
+  });
+};
```
Am I missing something or isn't this file redundant with 'redux/actions/statistics/index.js' below?

matanwerbner replied:
yes it is , it's removed by now

gailsteiger commented:
Kudos Matan. It works, and to the extent that I understand, looks good.

There are a couple of instances where I have needed to use local/session storage to 'maintain state' following page navigation. You can see that with Notifications and it will also be necessary with Toasts. Is this going to be a problem?

matanwerbner commented:
Thanks, 
Initial state is built in to redux, as you can see in the store - each reducer accepts his initial state, which often comes from local storage.

gailsteiger commented:
We should settle on a consistent way to handle test file naming and directory location. I'll have another look  at jest conventions. In most cases I don't have strong opinions but we should use the same convention.

gailsteiger commented:
Where would we update local storage when state changes?  For instance, the navigation drawer (if open) should look the same after navigation as before. The user can open/close the drawer or open a different accordion panel before the navigation. This will need to be reflected in local storage before the navigation.

matanwerbner commented:
side effects such as API calls, session storage or similar, should be handled in the action creators, please see how the API is called on redux/actions/statistics/index.js
it would be almost the same handling sessionStorage: make the side-effect, then dispatch the actual action.

gailsteiger commented:
Is webpack/assets/javascripts/react_app/actions/StatisticsChartActions.js being used. Likewise for API.getStatisticsData

matanwerbner commented:
good point, i'll remove those

gailsteiger commented:
Looks like webpack/assets/javascripts/react_app/stores/StatisticsStore.js isn't being used either

gailsteiger commented on webpack/assets/javascripts/react_app/redux/actions/statistics/index.js:
```diff
@@ -0,0 +1,18 @@
+      )
+    );
+  });
+};
```
Will the above replace 'webpack/assets/javascripts/react_app/actions/ServerActions.js' in the event that we completely replace flux with redux.

matanwerbner replied:
absolutley

ohadlevy commented on webpack/assets/javascripts/react_app/components/charts/spec/fixtures.js:
```diff
@@ -0,0 +1,14 @@
+  {
+    id: 'operatingsystem',
+    title: 'OS Distribution',
+    url: 'statistics/operatingsystem',
```
this spec does not include actually any data, is that on purpose?

matanwerbner replied:
Yes, and It was a dilema, i thought that the way components are currently built, data-specific tests may belong in the chartbox component, since the list has a job of rendering x panels for x items in the state.
tests of this component can be improved for sure, including a mount test to see if actions are fired properly and empty store gets proper values when given data.

matanwerbner commented:
made the reducer syntax more simple with swtich / case, removed "spec" libraries to make file structue more uniform, and removed "status" property from the state ( it is implicit ).

i'll be happy to answer any other suggestions, if not this PR is complete by me.

ohadlevy commented:
@matanwerbner can you please remove the statistics actions from webpack/assets/javascripts/react_app/actions/ServerActions.js ?
there are references also to the constant file

ohadlevy commented on webpack/assets/javascripts/react_app/redux/index.js:
```diff
@@ -0,0 +1,12 @@
+
+export default createStore(
+  reducer,
+  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
```
is this part of the PR?

matanwerbner replied:
yes, this enabled redux devtools on the browser
http://extension.remotedev.io/

ohadlevy commented on webpack/assets/javascripts/react_app/redux/actions/statistics/index.js:
```diff
@@ -0,0 +1,18 @@
+
+export const getStatisticsData = charts => dispatch => {
+  dispatch({ type: STATISTICS_DATA_REQUEST, payload: charts });
+  charts.forEach(c => {
```
nitpick: can you use chart vs c for more meaningful var names?

gailsteiger replied:
+1

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/statistics/index.js:
```diff
@@ -0,0 +1,29 @@
+  switch (action.type) {
+    case STATISTICS_DATA_REQUEST: return state.set('charts', payload);
+    case STATISTICS_DATA_SUCCESS:
+      return state.set('charts', state.charts.map(chart => chart.id === action.payload.id ?
```
nitpick: action.payload.id can simply be payload.id ?

ohadlevy commented on webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.js:
```diff
@@ -1,111 +1,49 @@
-      'onChange',
-      'onError']
-    );
+const getStatusFromChart = (chart) => {
```
@gailsteiger any concerns about this change?

gailsteiger replied:
OK

ohadlevy commented on webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.js:
```diff
@@ -1,111 +1,49 @@
-      'onChange',
-      'onError']
-    );
+const getStatusFromChart = (chart) => {
```
@matanwerbner also, why isnt this a function within the react class?

matanwerbner replied:
this is a utility function. i dont think it belongs in the component, since it is not react-related.
anyway we should refactor this to avoid rendering the "smart" chartbox component.
after refactoring this function will go away.
i think it can also be in a utilities file, along with the translation strings.

gailsteiger replied:
@matanwerbner could you elaborate on refactoring to avoid rendering the "smart" chartbox component?

matanwerbner replied:
i meant that i believe the list component should contain logic as to what chart it renders.
it should choose between rendering a chart, a loading panel,  or an error panel,  based on the item it is handling.
so the chart box might be not neccesary

gailsteiger replied:
Based on what you're saying, the `Loader` would not be necessary but I'm not sure about the `ChartBox`. 

Generally, what thoughts about the pros and cons of using components like the [`Loader`](https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/Loader.js) or [`Fade`](https://github.com/gailsteiger/foreman/blob/pf-react-toast/webpack/assets/javascripts/react_app/components/common/Fade.js) which wrap other components and manage part of the wrapped components' rendering logic as opposed to including the relevant logic in the wrapped components themselves?

matanwerbner replied:
i think wrapper components are great, it's a very common pattern in react and functional programming, in common usecases.
i feel that there are cases where it's not required, and adds a level of complexity to the code.
if we have common functionality that spans several components, it's worth considering making a wrapper (if normal solutions such as css dont work out)...

matanwerbner replied:
after looking at loader, i think i would prefer manually adding a loading component for components that require it. 
makes it easier to directly describe and understand the logic for showing/hiding the loader, rather then requiring status parameter

gailsteiger replied:
good point

ohadlevy commented:
@matanwerbner note webpack/assets/javascripts/react_app/constants.js still reference statistics.

matanwerbner commented:
@ohadlevy fixed.

ohadlevy commented on webpack/assets/javascripts/react_app/components/charts/StatisticsChartsList.spec.js:
```diff
@@ -0,0 +1,42 @@
+import { shallow } from 'enzyme';
+import StatisticsChartsList from './StatisticsChartsList';
+import { statisticsData } from './StatisticsChartsList.fixtures';
+
```
nitpick: extra line

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/statistics/statistics.fixtures.js:
```diff
@@ -0,0 +1,38 @@
+];
+export const response = {
+  id: 'operatingsystem',
+  data: [['redHat 3', 2]]
```
nitpick: redHat is not a valid brand, either write RHEL or RedHat :)

ohadlevy commented on webpack/assets/javascripts/react_app/redux/reducers/statistics/statistics.fixtures.js:
```diff
@@ -0,0 +1,38 @@
+});
+
+export const stateAfterSuccess = Immutable({
+  charts: request.map(c => ({
```
same nitpick about one chart variable names :)

ohadlevy commented:
@matanwerbner besides minor nitpicks big :+1:  for this change.

I hope @gailsteiger is around to double check me, but from my side :+1: to merge.

gailsteiger commented:
OK to merge. Super!

ohadlevy commented:
thanks @matanwerbner !
